linux/dm-sworndisk/src/lib.rs

```diff
@@ -3,36 +3,351 @@
 
 //! Dummy dm-sworndisk module.
 
-use kernel::prelude::*;
+#![feature(allocator_api)]
+
+mod bio;
+mod block_device;
+mod device_mapper;
+
+use bindings::thread::{spawn, JoinHandle};
+use core::{
+    ops::{Deref, DerefMut, Range},
+    sync::atomic::{AtomicBool, Ordering},
+};
+use crossbeam_queue::SegQueue;
+use kernel::{c_str, new_condvar, prelude::*, sync::CondVar};
+use sworndisk::{Arc, BlockId, BlockSet, BufMut, BufRef, Errno, Error, Mutex};
+
+use crate::bio::{Bio, BioOp, BioVec};
+use crate::block_device::{HostBlockDevice, BLOCK_SECTORS, BLOCK_SIZE};
+use crate::device_mapper::{Args, DmTargetOps, MapState, Target, TargetType};
 
 module! {
-    type: Dummy,
+    type: TargetManager,
     name: "dm_sworndisk",
     author: "Rust for Linux Contributors",
     description: "Rust dm_sworndisk module",
     license: "GPL",
 }
 
-struct Dummy;
+/// A struct used to manage a `TargetType`.
+///
+/// Register the type to kernel when `init` the module, and `unregister` it
+/// when `drop` (automatically).
+struct TargetManager(Pin<Box<TargetType>>);
 
-impl kernel::Module for Dummy {
+impl kernel::Module for TargetManager {
     fn init(_module: &'static ThisModule) -> Result<Self> {
         pr_info!("Rust dm_sworndisk module (init)\n");
 
-        test_rwlock();
-        test_weak();
-        test_thread();
+        // TODO: add a switch to enable those tests.
+        // test_rwlock();
+        // test_weak();
+        // test_thread();
 
-        Ok(Dummy)
+        let dm_sworndisk = Box::pin_init(TargetType::register::<DmSwornDisk>(
+            c_str!("sworndisk"),
+            [0, 0, 1],
+            0,
+        ))?;
+
+        Ok(TargetManager(dm_sworndisk))
     }
 }
 
-impl Drop for Dummy {
+impl Drop for TargetManager {
     fn drop(&mut self) {
         pr_info!("Rust dm_sworndisk module (exit)\n");
     }
 }
 
+/// A request queue, dispatching bios from device mapper to `RawDisk`.
+struct ReqQueue {
+    bios: Mutex<SegQueue<Bio>>,
+    // TODO: replace `RawDisk` with `SwornDisk`.
+    disk: RawDisk,
+    should_stop: AtomicBool,
+    new_bio: Pin<Box<CondVar>>,
```

Comment:
`new_bio` is a bad name. It sounds like it is a `Bio`, but it is not. Name it `new_bio_condvar` to make it more clear.